zerver/worker/queue_processors.py

```diff
@@ -616,108 +622,106 @@ def consume(self, event: Dict[str, Any]) -> None:
                     scheduled_timestamp=scheduled_timestamp,
                     mentioned_user_group_id=event.get("mentioned_user_group_id"),
                 )
-
-                self.ensure_timer()
+                if not already_pending:
+                    # If this was the first row, we have to wake the cv
+                    # from its timeout-less sleep.
+                    self.cv.notify()
             except IntegrityError:
                 logging.debug(
                     "ScheduledMessageNotificationEmail row could not be created. The message may have been deleted. Skipping event."
                 )
 
-    def ensure_timer(self) -> None:
-        # The caller is responsible for ensuring self.lock is held when it calls this.
-        if self.timer_event is not None:
-            return
+    def start(self) -> None:
+        self.stopping = False
+        self.worker_thread = threading.Thread(target=lambda: self.work())
+        self.worker_thread.start()
+        super().start()
 
-        self.timer_event = Timer(
-            self.TIMER_FREQUENCY, MissedMessageWorker.maybe_send_batched_emails, [self]
-        )
-        self.timer_event.start()
+    def work(self) -> None:
+        while True:
+            with self.cv:
+                if self.stopping:
+                    return
+                # Only set a timeout if we know there rows to
+                # aggregate; we will be notified on exit or on there
+                # being rows.
+                timeout: Optional[int] = None
+                if ScheduledMessageNotificationEmail.objects.exists():
+                    timeout = self.CHECK_FREQUENCY_SECONDS
+
+                # Wait until we have something in the queue, or we're
+                # stopping, or we time out on the stored
+                # ScheduledMessageNotificationEmail rows.  Ordering on
+                # the lambda criteria is important, so we save the
+                # database query if possible.
+                was_notified = self.cv.wait_for(
+                    lambda: self.stopping
+                    or (timeout is None and ScheduledMessageNotificationEmail.objects.exists()),
+                    timeout=timeout,
+                )
```

Comment:
Since we don't have much condition variable code, it might be worth spelling out what `wait_for` does in a few words -- it's not entirely obvious to me on looking at this how frequently we expect `ScheduledMessageNotificationEmail.objects.exists` to be called. Or maybe I just want the cases spelled out a bit more like you have above `already_pending`.

(Also, should we replace the lambda with a little function just to avoid the somewhat hard-to-read line wrapping?)

Reply:
Updated!